dsgrid/spark/functions.py

```diff
@@ -398,28 +454,74 @@ def prepare_timestamps_for_dataframe(timestamps: Iterable[datetime]) -> Iterable
-            if "PYTEST_VERSION" not in os.environ:
-                msg = f"cast_timestamp in read_csv can only be set in a test environment: {path=}"
-                raise Exception(msg)
-            # TODO: need a better way of guessing and setting the correct type.
-            df = df.withColumn("timestamp", F.col("timestamp").cast(TimestampType()))
-        dup_cols = [x for x in df.columns if x.endswith(".1")]
-        if dup_cols:
-            msg = f"Detected a duplicate column in the dataset: {dup_cols}"
-            raise DSGInvalidDimension(msg)
+    schema_file = _get_schema_file(path)
+    if schema_file is None:
+        csv_schema = None
+        mapping_schema = None
+    else:
+        try:
+            csv_schema = CsvSchema.from_file(schema_file)
+        except ValidationError:
+            logger.exception("Failed to load the schema file %s", schema_file)
+            msg = f"Invalid {schema_file=}"
+            raise DSGInvalidDataset(msg)
+        mapping_schema = decode_schema(csv_schema)
+    func = _read_csv_duckdb if use_duckdb() else _read_csv_spark
+    df = func(path, mapping_schema)
+    if mapping_schema is not None:
+        if set(df.columns).symmetric_difference(mapping_schema.keys()):
+            msg = (
+                f"Mismatch in CSV schema ({sorted(mapping_schema.keys())}) "
+                f"vs DataFrame columns ({df.columns})"
+            )
+            raise DSGInvalidDataset(msg)
+
+    if csv_schema is None:
         return df
-    return spark.read.csv(str(path), header=True, inferSchema=True)
+    return _rename_columns(df, get_column_renames(csv_schema))
+
+
+def _read_csv_spark(path: Path | str, schema: dict[str, str] | None) -> DataFrame:
+    spark = get_spark_session()
+    if schema is None:
+        return spark.read.csv(str(path), header=True, inferSchema=True)
+
+    schema_str = ",".join([f"{key} {val}" for key, val in schema.items()])
+    return spark.read.csv(str(path), header=True, schema=schema_str)
+
+
+def _read_csv_duckdb(path: Path | str, schema: dict[str, str] | None) -> DataFrame:
+    path_ = path if isinstance(path, Path) else Path(path)
+    if path_.is_dir():
+        path_str = str(path_) + "**/*.csv"
+    else:
+        path_str = str(path_)
+
+    spark = get_spark_session()
+    if schema is None:
+        return spark.read.csv(str(path), header=True)
+
+    dtypes = {k: duckdb.type(v) for k, v in schema.items()}
+    rel = duckdb.read_csv(path_str, header=True, dtype=dtypes)
+    columns = ",".join(schema.keys())
+    return spark.createDataFrame(rel.select(columns).to_df())
+
+
+def _get_schema_file(filename: Path | str) -> Path | None:
+    table_path = Path(filename)
+    for ext in (".json", ".json5"):
+        schema_file = table_path.with_name(f"{table_path.stem}_schema{ext}")
+        if schema_file.exists():
+            return schema_file
+    return None
+
+
+def _rename_columns(df: DataFrame, mapping: dict[str, str]) -> DataFrame:
```

Comment:
This code path is not called when reading with Spark. We can easily add it in this PR. Need to add a test as well. Also need to add it to the documentation.